projects/media-viewer/src/lib/annotations/comment-set/comment-set-render.service.ts

```diff
@@ -4,22 +4,25 @@ import { CommentComponent } from './comment/comment.component';
 @Injectable()
 export class CommentSetRenderService {
 
-  redrawComponents(commentComponents: CommentComponent[], height: number, rotate: number, zoom: number) {
-      let prevComment: CommentComponent;
-      this.sortComponents(commentComponents, height, rotate, zoom).forEach((comment: CommentComponent) => {
-        this.adjustIfOverlapping(comment, prevComment, zoom);
-        prevComment = comment;
-      });
-      prevComment = null;
-      commentComponents.reverse().forEach((comment: CommentComponent) => {
-        prevComment = this.makeSureWithinContainer(comment, prevComment, height);
-      });
+  pageHeights = [];
+
+  redrawComponents(commentComponents: CommentComponent[], pageHeights: any[], containerHeight: number, rotate: number, zoom: number) {
+    this.pageHeights = pageHeights;
+    let prevComment: CommentComponent;
+    this.sortComponents(commentComponents, rotate, zoom).forEach((comment: CommentComponent) => {
+      this.adjustIfOverlapping(comment, prevComment, zoom);
+      prevComment = comment;
+    });
+    prevComment = null;
+    commentComponents.reverse().forEach((comment: CommentComponent) => {
+      prevComment = this.makeSureWithinContainer(comment, prevComment, containerHeight);
+    });
```

Comment:
This method could probably be better:

```
redrawComponents(blah...) {
  this.pageHeights = pageHeights; // little bit of vomit here but we're good
  this.sortComponents(commentComponents, rotate, zoom); // sort method mutates 

  for (let i = 1; i < commentComponents.length; i++) {
    this.adjustIfOverlapping(commentComponents[i], commentComponents[i - 1], zoom);
  }
  
  for (let i = commentComponents.length - 2; i >= 0; i++) {
    this.makeSureWithinContainer(commentComponents[i], commentComponents[i + 1], zoom);
  }
```

I'm fairly confident if I knew what `adjustIfOverlapping` and `makeSureWithinContainer` it might be possible to make it better. There's lots of mutable objects here which feels a bit ugly, but it's also how CSS/Dom stuff is.

Main main objections here is `.forEach` - ain't no reason for that. Just use a `for (const derp of derpy) ` it's much easier to read and doesn't involve the overhead of a lambda.

Reply:
Just noticed that `makeSureWithinContainer` returns the `prevComment` so my example above may not be valid, even so, it seems that the prevComment variable should be avoidable.

Reply:
ill look into that, I believe pawel created it to push the comments up instead of making it scrollable.